### PR TITLE
issues: Add comments and actions for labels, milestones, titles and assignee changes

### LINKED ISSUE #3540: No email notification is sent when issue is assigned
- Gogs version (or commit ref): v0.9.71
- Git version: v2.1.4
- Operating system: Debian
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gogs.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant

## Description

When assigning an issue, it is valuable to notify all subscribed users. Perhaps also making this mutations (also label & milestone mutations) visible in the history would be nice.

### PR DESCRIPTION
**Pull request by hochhaus**
Fixes #3540

This is my first gogs change that is longer than a few lines. It is very likely that I missed something. I'll happily fix any issues that you find.

### CODE REVIEW

**Comment by Unknwon** [comment]
Thanks for the PR!

A quick review found you didn't use `.less` to generate `.css`.

**Comment by hochhaus** [comment]
Thanks for the first pass review. Fixed. Can you have another look as time permits?

**Comment by hochhaus** [comment]
@Unknwon Any chance you will have some review bandwidth this week? I have another change that I would like to make which builds upon this one. However, I don't was to go too far down the path until I can improve this code to fit well in gogs and meet your standards.

**Comment by Unknwon** [comment]
I'll get sometime tomorrow, sorry about waiting!

**Comment by hochhaus** [comment]
Thanks @Unknwon. I appreciate your time.

File: models/comment.go [hunk 1]
```diff
@@ -60,6 +69,24 @@ type Comment struct {
...
+
+	MilestoneID    int64      `xorm:"INDEX" json:"-"`
+	Milestone      *Milestone `xorm:"-" json:"-"`
+	OldMilestoneID int64      `xorm:"INDEX" json:"-"`
+	OldMilestone   *Milestone `xorm:"-" json:"-"`
+
+	AssigneeID    int64 `xorm:"INDEX" json:"-"`
+	Assignee      *User `xorm:"-" json:"-"`
+	OldAssigneeID int64 `xorm:"INDEX" json:"-"`
+	OldAssignee   *User `xorm:"-" json:"-"`
```
Comment (Unknwon) [inline models/comment.go]: Any reason we need INDEX for these new "*ID"s?

File: models/comment.go [hunk 2]
```diff
@@ -36,6 +36,15 @@ const (
...
 	COMMENT_TYPE_PULL_REF
+
+	// Labels
+	COMMENT_TYPE_LABEL
+	// Milestones
+	COMMENT_TYPE_MILESTONE
+	// Title
+	COMMENT_TYPE_TITLE
+	// Assignee
+	COMMENT_TYPE_ASSIGNEE
```
Comment (Unknwon) [inline models/comment.go]: Add `_CHANGE` suffix to these four new types is more accurate.

File: models/action.go [hunk 3]
```diff
@@ -51,6 +52,10 @@ const (
 	ACTION_MIRROR_SYNC_PUSH                          // 20
 	ACTION_MIRROR_SYNC_CREATE                        // 21
 	ACTION_MIRROR_SYNC_DELETE                        // 22
+	ACTION_LABEL_ISSUE                               // 23
+	ACTION_MILESTONE_ISSUE                           // 24
+	ACTION_TITLE_ISSUE                               // 25
+	ACTION_ASSIGNEE_ISSUE                            // 26
```
Comment (Unknwon) [inline models/action.go]: `ACTION_ISSUE_***_CHANGE` feels more accurate.

File: models/comment.go [hunk 4]
```diff
@@ -273,12 +409,32 @@ func createComment(e *xorm.Session, opts *CreateCommentOptions) (_ *Comment, err
...
+		act.OpType = ACTION_ASSIGNEE_ISSUE
+		act.Content = fmt.Sprintf("%d|%d", opts.Issue.Index, comment.ID)
 	}
 
 	if _, err = e.Exec("UPDATE `issue` SET updated_unix = ? WHERE id = ?", time.Now().Unix(), opts.Issue.ID); err != nil {
 		return nil, fmt.Errorf("update issue 'updated_unix': %v", err)
 	}
 
+	if err := comment.loadAttributes(e); err != nil {
+		return nil, err
```
Comment (Unknwon) [inline models/comment.go]: `fmt.Errorf("loadAttributes: %v", err)`

File: pkg/template/template.go [hunk 5]
```diff
@@ -276,6 +276,14 @@ func ActionIcon(opType int) string {
 		return "repo-forked"
 	case 20, 21, 22: // Mirror sync
 		return "repo-clone"
+	case 23:
```
Comment (Unknwon) [inline pkg/template/template.go]: Add a short description comment as above cases.

**Comment by hochhaus** [comment]
Example of CSS/UI change. Just adding inline updates (similar to github) for changes to labels, milestones, titles and assignees.

gogs

File: models/comment.go (hunk 1 above)
Comment (hochhaus) [reply models/comment.go]: No. I will remove.

File: models/comment.go (hunk 2 above)
Comment (hochhaus) [reply models/comment.go]: Done.

File: models/action.go (hunk 3 above)
Comment (hochhaus) [reply models/action.go]: Done

File: models/comment.go (hunk 4 above)
Comment (hochhaus) [reply models/comment.go]: Done

File: pkg/template/template.go (hunk 5 above)
Comment (hochhaus) [reply pkg/template/template.go]: Done

**Comment by hochhaus** [comment]
@Unknwon Thanks for the review. I addressed your comments and added a screenshot. Could you please take another look as you have time?

**Comment by Unknwon** [comment]
Sorry forgot to mention would you mind provide a screenshot of dashboard feeds?

**Comment by Unknwon** [comment]
Overall, it is ready to merge, thanks for the good work!

I just need to do a final round of careful review.

File: templates/user/dashboard/feeds.tmpl [hunk 6]
```diff
@@ -86,6 +98,18 @@
 						<p class="text light grey has-emoji">{{index .GetIssueInfos 1}}</p>
 					{{else if (or (or (eq .GetOpType 12) (eq .GetOpType 13)) (or (eq .GetOpType 14) (eq .GetOpType 15)))}}
 						<span class="text truncate issue title has-emoji">{{.GetIssueTitle}}</span>
+					{{else if eq .GetOpType 23}}
+						<span class="text truncate issue title has-emoji">{{.GetIssueTitle}}</span>
+            <p class="text light grey has-emoji">{{if .HasIssueRemovedLabel}}<a class="ui label" href="{{.GetRepoLink}}/issues?labels={{.GetIssueRemovedLabelID}}" style="color: {{.GetIssueRemovedLabelForegroundColor}}; background-color: {{.GetIssueRemovedLabelColor}}; text-decoration: line-through">{{.GetIssueRemovedLabelName | Sanitize}}</a>{{end}} {{if .HasIssueAddedLabel}}<a class="ui label" href="{{.GetRepoLink}}/issues?labels={{.GetIssueAddedLabelID}}" style="color: {{.GetIssueAddedLabelForegroundColor}}; background-color: {{.GetIssueAddedLabelColor}}">{{.GetIssueAddedLabelName | Sanitize}}</a>{{end}}</p>
```
Comment (Unknwon) [inline templates/user/dashboard/feeds.tmpl]: The only thing I concerned about is in this one single feed, could cause multiple DB queries. How about save the information (i.e. label ID, color, name) to the at creation of the feed (`Action`, use `|` to split). This way, we could basically avoid most of DB queries for displaying the feeds.

The feed content is meant to/could be in the state of immutable, so store the instant state of information at creation sounds better to me.

What do you think?

**Comment by hochhaus** [comment]
Thanks @Unknwon!

Here is a screenshot of the dashboard:
dashboard

I agree with your performance comment about the dashboard. I will send a change caching more of the required state to avoid extra database queries.

**Comment by hochhaus** [comment]
@unknwon Sorry that I dropped this change. Can you share your plans? If helpful, I can pick this change up again and make the required updates to minimize database queries if you are interested.

**Comment by unknwon** [comment]
FYI... The close of PR wasn't authorized by me, it's made by GitHub automatically, see my comments: https://github.com/gogs/gogs/pull/4551#issuecomment-546464935

---

I do have a wish to merge in this PR with comments I've made in last round of review.

**Pull request closed without merging**